src/backend/parser/parse_partition.c

```
@@ -1124,12 +1126,20 @@ make_partition_rules(CreateStmtContext *cxt, CreateStmt *stmt,
 					pEq = (Node *) nt;
 				}
 				else
+				{
+					Oid opcOid = lfirst_oid(lc_opclass);
+					Oid opcFamily = get_opclass_family(opcOid);
+					Oid opcInputType = get_opclass_input_type(opcOid);
+					Oid keyOp = get_opfamily_member(opcFamily, opcInputType, opcInputType, BTEqualStrategyNumber);
+					char *keyOpNamespace = get_namespace_name(get_opnamespace(keyOp));
+
 					/* equality expression: column = value */
-					pEq = (Node *) makeSimpleA_Expr(AEXPR_OP,
-													"=",
-													(Node *) pCRef,
-													(Node *) pValConst,
-													-1 /* position */ );
+					pEq = (Node *) makeA_Expr(AEXPR_OP,
+											  list_make2(makeString(keyOpNamespace), makeString("=")),
```

Is it possible `keyOpNamespace` got `NULL`?

---

Seems we need to error out when `keyOp` is Invalid.

---

Here we find the op's oid, and use it to search a namespace, and finally use the namespace search again to build an expr. Do I understand correctly?

Is there more straightforward way?

---

Since we're using a qualified operator name, we should update here as well. Though the `deparse_partition_rule` function seems for debugging?

https://github.com/greenplum-db/gpdb/blob/d11fb106e8ea732b0fd72f938e18ce142476dba3/src/backend/parser/parse_partition.c#L1604

---

> Is it possible keyOpNamespace got NULL?
> Seems we need to error out when keyOp is Invalid.

Yes, it seems we do need a check on invalid operator Oid.  Added invalid Oid check and added test coverage for it.

> Here we find the op's oid, and use it to search a namespace, and finally use the namespace search again to build an expr. Do I understand correctly?
>
> Is there more straightforward way?

Yes, that is correct.  It is indeed not very optimal unfortunately but this seems to be the most correct way to handle the issue.  I think this is the only spot where we have the distribution key's operator class so that we can get the correct equality operator for the custom type.

> Since we're using a qualified operator name, we should update here as well. Though the deparse_partition_rule function seems for debugging?

Interesting...  I think `deparse_partition_rule` is possibly dead code.  I see the output string being used in GPDB 5X but seems like GPDB 6X removed the logic?  Either way, I fixed the line as mentioned.  We can determine if the code needs to be removed later in a separate PR.

---

When the code reach Line 1143: `pEq = (Node *) makeA_Expr(AEXPR_OP,` (use text to search), there is a concurrent transaction running `alter schema xxx rename`. Will still lead this to not-correct bindings.

Seems an unsolvable issue on 6X.

The above corner case should be very rare.

---

> When the code reach Line 1143: `pEq = (Node *) makeA_Expr(AEXPR_OP,` (use text to search), there is a concurrent transaction running `alter schema xxx rename`. Will still lead this to not-correct bindings.
> 
> Seems an unsolvable issue on 6X.
> 
> The above corner case should be very rare.

Unfortunately yes; but at least the problem is shifted to a very unlikely scenario with an easy workaround (rerun the CREATE TABLE with the new schema name).  As it is today without this fix, the user would have to drop the entire partition table prior to gpupgrade (and has issues with pg_dump/pg_restore and probably gpbackup/gprestore too) due to the search_path dependency.